Common area

aravindakidambi opened the pull request:
Implement the reconciller for the ClusterSet CRD. The ClusterSet CRD depends on the ClusterClaim CRD, although
the two can be configured in any order, the ClusterSet CRD will perform validations to make sure it will be processed only after the appropriate ClusterClaim CRDs are present.

Leader ClusterSet controller - Currently creates access to the local namespace using the Common Area abstraction

```
 Leader ClusterSet Controller
            |
 InboundLeaderClusterManager
            |
       Common Area
  (access to local cluster
   namespace)
```

Member ClusterSet controller - read the leader list from spec and mantain a connection to each of them using the
controller-runtime library. This connection is abstracted as the common area of the leader cluster namespace for the
member to write to and listen for monitor updates

```
 Member ClusterSet Controller
            |
   Remote Cluster Manager
            | 1:N
      Remote Cluster
            |
      Common Area
  (access to remote cluster
   namespace)
```

luolanzone commented on multicluster/apis/multicluster/v1alpha1/zz_generated.deepcopy.go:
```diff
@@ -15,15 +16,13 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 */
-
-// Code generated by controller-gen. DO NOT EDIT.
+// Code generated by deepcopy-gen. DO NOT EDIT.
```
Can you point out which change involved this? I didn't see any change on Makefile, I suppose we should keep this consistent. by default, kubebuild use controller-gen to do this code auto generation.

aravindakidambi replied:
Im not sure, its not there anymore now after I rebase to tot and recompile

luolanzone commented on multicluster/config/manager/controller_manager_config.yaml:
```diff
@@ -9,3 +9,4 @@ webhook:
 leaderElection:
   leaderElect: true
   resourceName: 6536456a.crd.antrea.io
+isMcsLeader: true
```
I would prefer `isLeader` which is simpler.

jianjuns replied:
At least change Mcs to MCS (we use all upper cases for abbrs).

luolanzone commented on multicluster/config/rbac/secret_viewer_role.yaml:
```diff
@@ -0,0 +1,14 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: antrea-multicluster-secret-viewer-role
```
you need to remove `antrea-multicluster-` because it will be added automatically by kustomize, otherwise you will see name like `antrea-multicluster-antrea-multicluster-secret-viewer-role` in `multi-cluster.yaml `

luolanzone commented on multicluster/config/rbac/secret_viewer_role_binding.yaml:
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: antrea-multicluster-secret-viewer-rolebinding
```
ditto

luolanzone commented on multicluster/config/samples/multicluster_v1alpha1_clusterclaim.yaml:
```diff
@@ -1,7 +1,31 @@
+name: clusterSet.k8s.io
+value: test-clusterset
+
+#apiVersion: multicluster.crd.antrea.io/v1alpha1
```
I suppose these comments should be removed.

luolanzone commented on multicluster/config/samples/multicluster_v1alpha1_clusterset.yaml:
```diff
@@ -1,7 +1,35 @@
+#      # secret: "" # This is NA for inbound mode
+#      # server: "" # This is NA for inbound mode
+#      serviceAccount: "clusterset-member-1"
+#  namespace: member-ns-one
```
ditto

luolanzone commented on multicluster/config/samples/multicluster_v1alpha1_clusterclaim.yaml:
```diff
@@ -1,7 +1,31 @@
+#  name: member1clustersetid
+#  namespace: kube-system
+#name: clusterSet.k8s.io
+#value: test-clusterset
```
newline is missing here.

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -47,9 +70,48 @@ type ClusterSetReconciler struct {
+	if err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+	if errors.IsNotFound(err) {
```
you can make the if condition like below to avoid check the err type twice.
```go
if err != nil {
  if !errors.IsNotFound(err){
    return ctrl.Result{}, err
  }
  // other checks
}
```

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -47,9 +70,48 @@ type ClusterSetReconciler struct {
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *ClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	_ = r.Log.WithValues("ClusterSet", req.Namespace)
```
I suppose we can use klog here directly instead of creating a new log and new package dependency?

abhiraut replied:
+1.. see usage in antrea codebase

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -17,20 +17,43 @@ limitations under the License.
+	clusterID        common.ClusterID
+
+	// These fields are only applicable on leader cluster
+
```
remove extra empty line

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -17,20 +17,43 @@ limitations under the License.
+	inboundLeaderMgr internal.InboundLeaderClusterManager
+
+	// These fields are only applicable on member cluster
+
```
ditto

luolanzone commented on multicluster/apis/multicluster/v1alpha1/zz_generated.deepcopy.go:
```diff
@@ -44,14 +44,6 @@ func (in *ClusterClaim) DeepCopy() *ClusterClaim {
 }
 
-// DeepCopyObject is an autogenerated deepcopy function, copying the receiver, creating a new runtime.Object.
-func (in *ClusterClaim) DeepCopyObject() runtime.Object {
```
DeepCopyObject() is required, otherwise you will get similar error like below, so I am afraid you need to revert the change to this file.
```
cannot use clusterClaimList (variable of type *v1alpha1.ClusterClaimList) as client.ObjectList value in argument to r.List: missing method DeepCopyObject
```

aravindakidambi replied:
This should be fixed when we removed controller-gen to generate the deepcopy.go file

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+func (r *ClusterSetReconciler) validateLocalClusterClaim(clusterSet *multiclusterv1alpha1.ClusterSet) error {
+	// read the cluster claim in the cluster
+
```
extra empty line

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	wellKnownClusterClaimIDExist := false
+	for _, clusterClaim := range clusterClaimList.Items {
+		r.Log.Info("Processing clusterclaim", "name", clusterClaim.Name, "value", clusterClaim.Value)
+		if strings.Compare(clusterClaim.Name, multiclusterv1alpha1.WellKnownClusterClaimClusterSet) == 0 {
```
it's unnecessary to call `strings.Compare`, you can just use `clusterClaim.Name == multiclusterv1alpha1.WellKnownClusterClaimClusterSet`

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		if strings.Compare(clusterClaim.Name, multiclusterv1alpha1.WellKnownClusterClaimClusterSet) == 0 {
+			wellKnownClusterSetClaimIDExist = true
+			clusterSetClaimID = clusterClaim.Value
+		} else if strings.Compare(clusterClaim.Name, multiclusterv1alpha1.WellKnownClusterClaimID) == 0 {
```
ditto

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	wellKnownClusterSetClaimIDExist := false
+	var clusterClaimID string
+	wellKnownClusterClaimIDExist := false
+	for _, clusterClaim := range clusterClaimList.Items {
```
I didn't get this check, should this kind of logic being in clusterclaim_controller.go?

aravindakidambi replied:
To capture the offline conversation:
The idea is that we allow the customer to create ClusterSet CRD before creating the ClusterClaim CRDs and the reconciler will keep retrying processing the ClusterSet CRD until a valid ClusterClaim CRD is created.

Another way is: maybe move this validation to ClusterClaim controller, but since a valid CRD needs to process the ClusterSet, the current way maybe better.

A third way maybe to move this into a validator which may prevent ClusterSet CRD from being created, this may require customer to retry configuring the ClusterSet CRD. I think the current approach maybe better, so the customer can see the errors and fix it.

abhiraut replied:
perhaps for demo purposes this is okay..but it might be better to refactor this code when we open up a PR against feature branch by choosing one of the below approach
- maintain an indexer of the clusterSets referencing the claim and when the claim event occurs, trigger the reconciliation of the referencing clusterSets? OR
- Validate in webhook to ensure that the claim exists

abhiraut replied:
any thoughts on this?

luolanzone commented on multicluster/controllers/multicluster/internal/inbound_leader_manager.go:
```diff
@@ -0,0 +1,85 @@
+// Get namespace backing this common area
+func (i *inboundLeaderClusterManager) GetNamespace() string {
+	return i.namespace
+}
```
missing newline.

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+
+func (r *remoteCluster) IsConnected() bool {
+	return r.connected
+}
```
ditto

luolanzone commented on multicluster/main.go:
```diff
@@ -161,7 +173,7 @@ func main() {
 	}
 
-	setupLog.Info("starting manager")
+	setupLog.Info("starting modified manager")
```
didn't get the point of adding`modified`.

aravindakidambi replied:
I was testing this, will remove it.

luolanzone commented on multicluster/member_account.yaml:
```diff
@@ -0,0 +1,137 @@
+apiVersion: v1
```
I suppose this is for you local test purpose? the whole file should be removed.

luolanzone commented on multicluster/member_secret.yaml:
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
```
ditto

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	r.clusterID = common.ClusterID(clusterClaimID)
+	r.clusterSetID = common.ClusterSetID(clusterSetClaimID)
+	r.clusterSetConfig = clusterSet.DeepCopy()
```
I thought there will be a replica of ClusterSet in member cluster as CR, does this mean it's only in-memory object? how a  user can tell a member cluster which Clusterset it belows to?

abhiraut replied:
could you comment on this one?

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+			return fmt.Errorf("cluster not defined as leader in cluster set")
+		}
+		// also validate the namespace is the same
+		if strings.Compare(clusterSet.Spec.Namespace, clusterSet.GetNamespace()) != 0 {
```
ditto

abhiraut replied:
maybe move this check before the leader check .. no need to loop through leaders if the namespace don't match

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	configExists := false
+	if r.IsMcsLeader {
+		for _, leader := range clusterSet.Spec.Leaders {
+			if strings.Compare(clusterClaimID, leader.ClusterID) == 0 {
```
ditto

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	defer r.mutex.Unlock()
+
+	if r.inboundLeaderMgr == nil {
+		r.inboundLeaderMgr = internal.NewInboundLeaderClusterManager(r.Client, r.clusterID, clusterSet.GetNamespace(), r.Log)
```
Could you help to check https://github.com/abhiraut/antrea/blob/63d4af9be579bfb9f09b55a354e541a76f13b778/multicluster/controllers/multicluster/resourceexport_controller.go#L56 or refer to Antrea code to provide a method like `NewClusterSetReconciler` to do initialization in one place?

aravindakidambi replied:
Ill add a todo and get to this in the next set of changes

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	defer r.mutex.Unlock()
+
+	if r.remoteClusterMgr == nil {
+		r.remoteClusterMgr = internal.NewRemoteClusterManager(r.clusterSetID, r.Log, r.clusterID)
```
ditto

luolanzone replied:
I got `NewRemoteClusterManager not declared by package internal` error locally. did you miss something?
Could you run `make docker-build` and verify your code locally?

aravindakidambi replied:
Yeah somehow it was missed, I added it back now.

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		}()
+	}
+
+	if r.remoteClusterMgr == nil {
```
duplicate check `r.remoteClusterMgr == nil`, what's this for?

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+			var err error
+			err = r.remoteClusterMgr.Start()
+			if err != nil {
+				r.Log.Error(err, "Error starting remote remote cluster manager")
```
duplicate `remote`

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		close(ch)
+	}()
+
+	for _, addedLeader := range addedLeaders {
```
does this mean a member cluster will talk to multiple leaders? how does common area reflect this? I didn't quite get this part, Could you add more information on PR to tell the whole scenario? thanks.

aravindakidambi replied:
Yes that is right, as we discussed offline, I also added some comments, hopefully that will help.

abhiraut replied:
where can i find those comments?

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+		for _, memberAnnounce := range memberAnnounceList.Items {
+			if strings.Compare(memberAnnounce.ClusterID, string(r.remoteClusterManager.GetLocalClusterID())) == 0 {
+				localClusterMemberAnnounceExist = true
+				localClusterMemberAnnounce = memberAnnounce
```
I think you can break here once we found a matched memberAnnounce

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+		}
+	} else {
+		// create
+		localClusterMemberAnnounce.ClusterID = string(r.remoteClusterManager.GetLocalClusterID())
```
I am wondering which leader will be for this member cluster? I didn't see you set `LeaderClusterID`.

aravindakidambi replied:
Added a comment

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		return nil, err
+	}
+	log.Info("found", "secret", secretName, "crtData", crtData, "token", token)
```
I think critical data should not be in log.

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+
+var (
+	// Use a raw client to get member cluster secrets from antrea-plus-system only.
+	secretCient client.Client
```
this should be `secretClient`

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+func getSecretCACrtAndToken(namespace string, secretName string) ([]byte, []byte, error) {
+	var err error
+	mutex.Lock()
+	if secretCient == nil {
```
ditto

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+	var err error
+	mutex.Lock()
+	if secretCient == nil {
+		secretCient, err = client.New(ctrl.GetConfigOrDie(), client.Options{})
```
ditto

abhiraut commented on multicluster/config/multi-cluster.yaml:
```diff
@@ -1350,6 +1350,23 @@ metadata:
+metadata:
+  labels:
+    app: antrea
+  name: antrea-multicluster-antrea-multicluster-secret-viewer-role
```
typo?

abhiraut replied:
`antrea-multicluster` is repeated, probably need to remove the prefix from yamls?

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster_manager.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+type RemoteClusterManager interface {
+	// Start RemoteClusterManager on a go routine which blocks.
```
looks like this is not a completed sentence?

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster_manager.go:
```diff
@@ -0,0 +1,146 @@
+	GetLocalClusterID() common.ClusterID
+}
+
+// The manager manages all leaders defined in the cluster and manages common area access into them
```
`defined in the cluster` should be `defined in the cluster set`?

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	removedLeaders = currentLeaders
+	for _, remoteCluster := range removedLeaders {
+		removeErr := r.remoteClusterMgr.RemoveRemoteCluster(remoteCluster)
```
seems `RemoveRemoteCluster` only add an remove event to channel instead of remove it, it always return `nil` , I don't think you can get a real error here.

abhiraut commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -17,20 +17,43 @@ limitations under the License.
 package multicluster
```
i tend to think that it might be better for us to organize the code as follows:
```
// Leader cluster code here
 multicluster/leader/controllers/*
// Member cluster code here
 multicluster/member/controllers/*
// Common code here
 multicluster/member/utils(or common)/*
```

because much of the code here is `if leader .. else`

aravindakidambi replied:
Sure, I will make this change in a subsequent patch

abhiraut commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if r.IsMcsLeader {
+		for _, leader := range clusterSet.Spec.Leaders {
+			if strings.Compare(clusterClaimID, leader.ClusterID) == 0 {
+				configExists = true
```
no need to further loop; add a `break` here

jianjuns commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -17,20 +17,43 @@ limitations under the License.
 package multicluster
 
 import (
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
```
We group imports from different sources, so it should be:

```
	"context"
	"fmt"
	"strings"
	"sync"

	"github.com/go-logr/logr"
	"go.uber.org/multierr"
	"k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/runtime"
	ctrl "sigs.k8s.io/controller-runtime"
	"sigs.k8s.io/controller-runtime/pkg/client"

	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
	"antrea.io/antrea/multicluster/controllers/multicluster/common"
	"antrea.io/antrea/multicluster/controllers/multicluster/internal"
```

abhiraut replied:
may want to address this comment when you rebase.. in this file and in others too

```
// go std lib imports

// third party imports

// project imports
```

jianjuns commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+type LeaderElector interface {
+	// StartLeaderElection the leader elector, returns a cancelFunc to invoke when it must be stopped
```
starts the leader elector?

jianjuns commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,75 @@
+func (m *remoteClusterManager) DoLeaderElection() {
+	log := m.log.WithName("LeaderElector")
+
+	// We have written member announce atleast once to all remote clusters, pick one randomly if it is connected
```
atleast -> at least

jianjuns replied:
, pick -> . Pick

jianjuns commented on multicluster/controllers/multicluster/internal/inbound_leader_manager.go:
```diff
@@ -0,0 +1,85 @@
+)
+
+/**
+ *  This file contains implementation for the leader cluster to manage access to common area in the inbound mode
```
I suggest to remove "inbound" and "inbound mode" from names and comments, until we decide to support "outbound" (and we should explain what does inbound/outbound mean somewhere), as it is confusing for readers have no context.

jianjuns commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFunc()
+	if localClusterMemberAnnounceExist {
+		if r.remoteClusterManager.GetElectedLeaderClusterID() != common.INVALID_CLUSTER_ID {
```
I think better to add more comments somewhere to explain how leader election is done. Not easy to figure out just from code.

jianjuns commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,300 @@
+type RemoteCluster interface {
+	common.CommonArea
+
+	Start() (context.CancelFunc, error)
```
Probably add some comments to describe what RemoteCluster will do in the started routine.

jianjuns reviewed (commented):
Before merge, please add more information about the change in the commit message (and better in the PR description too), so people can know what it is about from commit history.

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -47,9 +67,50 @@ type ClusterSetReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
```
you can remove default comments added by kubebuilder and add more description about the logic of clusterset reconciler.

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -17,20 +17,40 @@ limitations under the License.
+	inboundLeaderMgr internal.LocalClusterManager
+
+	// These fields are only applicable on member cluster
+	remoteClusterMgr internal.RemoteClusterManager
```
should the first letter of `inboundLeaderMgr` and `remoteClusterMgr` be upper case so they can be reused in other controllers? otherwise I didn't see how common area can be used without creating a new `RemoteClusterManager`

abhiraut replied:
btw also need to drop inbound/outbound references before we fully commit to outbound mode..

abhiraut commented:
please also add some description on the PR along with rebasing the conflicts

abhiraut commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -17,20 +17,40 @@ limitations under the License.
-	Scheme *runtime.Scheme
+	mutex    sync.Mutex
+	Scheme   *runtime.Scheme
+	Log      logr.Logger
```
i feel like we should use klog everywhere for logging. see usage in this PR https://github.com/abhiraut/antrea/pull/22/files or in general in antrea codebase

aravindakidambi replied:
I tried to do it, but if we are just doing .info or .error etc it works well, but if I want to set some common parameters to be printed for a log instance or some context

klogr.New().WithName("controllers"), 

it gives me logr.Logger instance.

abhiraut commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -47,9 +67,50 @@ type ClusterSetReconciler struct {
+	}
 
-	// your logic here
+	// it is okay to be both a leader and member
```
this comment feels like it is misplaced

abhiraut commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +121,176 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		}
+	}
+
+	ch := make(chan error)
```
```suggestion
	errCh := make(chan error)
```

sounds more descriptive

abhiraut commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,105 @@
+ * from the list of leader clusters in the ClusterSet spec.
+ *
+ * Leader Election picks an elected-leader among "connected" leaders
+ * by choosing one randomly
```
```suggestion
 * by choosing one randomly.
```

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -60,3 +121,176 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		return err
+	}
+	if len(clusterClaimList.Items) == 0 {
+		return fmt.Errorf("cluster claim not configured for the cluster")
```
s/cluster claim not configured for the cluster/cluster claim is not configured for the cluster/

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,320 @@
+func (r *remoteCluster) SendMemberAnnounce() error {
+	log := r.log.WithName("MemberAnnounce" + string(r.ClusterID))
+
+	log.Info("Writing member announce", "cluster", r.GetClusterID())
```
I was trying common area locally, I think this line can be removed since `SendMemberAnnounce()`  will be called every 5s.

aravindakidambi replied:
Sure, will remove it, this was useful for debugging, so maybe I will keep it until the demos are done?

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,320 @@
+}
+
+func (r *remoteCluster) SendMemberAnnounce() error {
+	log := r.log.WithName("MemberAnnounce" + string(r.ClusterID))
```
I think `MemberAnnounce` is enough since you already added ClusterID info here https://github.com/abhiraut/antrea/pull/18/files#diff-ea4c830887dcd27f542c4d4cfddaacfba7041173c80b544d63e415ea4e566127R86

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,320 @@
+func (r *remoteCluster) SendMemberAnnounce() error {
+	log := r.log.WithName("MemberAnnounce" + string(r.ClusterID))
+
+	log.Info("Writing member announce", "cluster", r.GetClusterID())
```
this log looks unnecessary since `r.GetClusterID()` is the same as `r.ClusterID`, which means duplicates.

aravindakidambi replied:
The first one (line 186) doesnt log anything, so Ill keep this but remove clusterID

luolanzone commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,320 @@
+		}
+		localClusterMemberAnnounce.Annotations["touch-ts"] = time.Now().String()
+		// do an update
+		if err := r.Update(ctx, &localClusterMemberAnnounce, &client.UpdateOptions{}); err != nil {
```
you should check existing memberannounce and do comparison before make a real update call, I see the existing memberannounce being updated every 5s in leader cluster with only `touch-ts` updated. btw, it's better to change `touch-ts` to a more clear name like `touch-timestamp`.

aravindakidambi replied:
I wanted to update it every interval. This way the leader can check if there are 3 consecutive missed updates, then, it will mark the cluster as disconnected. Because the leader cannot distinguish between when a member is disconnected vs, when it is just not writing member-announce because there is no change.

abhiraut replied:
im afraid of the number of updates that will be done every 5s just for timestamps

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -47,9 +67,50 @@ type ClusterSetReconciler struct {
+			return ctrl.Result{}, err
+		}
+	} else {
+		r.createOrUpdateMultiClusterSetOnLeaderCluster(clusterSet)
```
looks like you only create `InboundLeaderMgr` in leader cluster, then no manager can represent a local cluster, I suppose you need a new field eg:`localClusterMgr` in `ClusterSetReconciler` to represent a local member cluster itself, so other controllers can reuse it.

aravindakidambi replied:
I dint understand, can you please elaborate this for me?

luolanzone replied:
I was working on integrate serviceexport controller with common area, but I found there is no way to get local member cluster ID via common area, I suppose it can be initialized by ClusterSet controller after ClusterClaim is added in local cluster?

aravindakidambi replied:
Got it, instead of adding it in the common area which is required to access every leader, I added it in the RemoteClusterManager
	// GetLocalClusterID returns local cluster ID
	GetLocalClusterID() common.ClusterID

Can you use that one?

luolanzone replied:
OK, I can try use it, but it feel a little strange to get a local cluster id from a `RemoteClusterManager`

aravindakidambi committed:
ClusterSet controller and Common Area interface

ClusterSet controller - implement the reconciller for the ClusterSet
CRD. The ClusterSet CRD depends on the ClusterClaim CRD, although
the two can be configured in any order, the ClusterSet CRD will
perform validations to make sure it will be processed only after
the appropriate ClusterClaim CRDs are present.

Leader ClusterSet controller - Currently creates access to the
local namespace using the Common Area abstraction

 Leader ClusterSet Controller
            |
 InboundLeaderClusterManager
            |
       Common Area
  (access to local cluster
   namespace)

Member ClusterSet controller - read the leader list from spec
and mantain a connection to each of them using the controller-runtime
library. This connection is abstracted as the common area of the leader
cluster namespace for the member to write to and listen for
monitor updates

 Member ClusterSet Controller
            |
   Remote Cluster Manager
            | 1:N
      Remote Cluster
            |
      Common Area
  (access to remote cluster
   namespace)

abhiraut commented on multicluster/build/yamls/antrea-multicluster-leader-only.yml:
```diff
@@ -1389,6 +1389,23 @@ rules:
+metadata:
+  labels:
+    app: antrea
+  name: antrea-multicluster-secret-viewer-role
```
should this be named `antrea-multicluster-leader-secret-viewer-role` similarly append `member` for member roles

abhiraut commented on multicluster/cmd/multicluster-controller/controller.go:
```diff
@@ -137,7 +141,7 @@ func run(o *Options) error {
 	}
 
-	setupLog.Info("starting manager")
+	setupLog.Info("starting modified manager")
```
maybe a more descriptive log?

aravindakidambi replied:
Removed it sorry, it was just for some testing

abhiraut commented on multicluster/cmd/multicluster-controller/options.go:
```diff
@@ -15,10 +15,12 @@
-	"fmt"
-
 	mcsv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"fmt"
```
i think the import order here will be as follows

```
import (
 	"fmt"
 	"io/ioutil"

 	"github.com/spf13/pflag"
 	"gopkg.in/yaml.v2"
 	"k8s.io/klog/v2"
 	ctrl "sigs.k8s.io/controller-runtime"

        mcsv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
 )
```

abhiraut replied:
this is still not well formatted

abhiraut commented on multicluster/cmd/multicluster-controller/options.go:
```diff
@@ -41,12 +45,31 @@ func (o *Options) complete(args []string) error {
 	var err error
 	options := ctrl.Options{Scheme: scheme}
 	ctrlConfig := &mcsv1alpha1.MultiClusterConfig{}
+	klog.Info("Config file is ", o.configFile)
```
i am assuming all these debug logs will be removed .. maybe we can add a TODO so as not to forget .. for after demo

abhiraut commented on multicluster/cmd/multicluster-controller/options.go:
```diff
@@ -41,12 +45,31 @@ func (o *Options) complete(args []string) error {
+		if err != nil {
+			return err
+		}
+		o.leader = ctrlConfig.Leader
```
`options` at L50 does not load these options? 

if not, then i would suggest a new method  `o.loadConfigFromFile()`

aravindakidambi replied:
Sure, added a todo to resolve this in the next patch

abhiraut commented on multicluster/cmd/multicluster-controller/options.go:
```diff
@@ -27,6 +29,8 @@ type Options struct {
 	configFile     string
 	SelfSignedCert bool
 	options        ctrl.Options
+	leader         bool
```
i think we should be able to directly access the leader value from `Options.options` instead of storing it in this struct

aravindakidambi replied:
Options.options is of type ctrl.Options, that wont have leader/member flags. It only as what controller-runtime library needs to run its logic right?

abhiraut commented on multicluster/controllers/multicluster/internal/local_cluster_manager.go:
```diff
@@ -0,0 +1,85 @@
+	stopFunc context.CancelFunc
+}
+
+func NewInboundLeaderClusterManager(client client.Client, clusterID common.ClusterID, namespace string, log logr.Logger) LocalClusterManager {
```
```suggestion
func NewLocalClusterManager(client client.Client, clusterID common.ClusterID, namespace string, log logr.Logger) LocalClusterManager {
```

abhiraut commented on multicluster/controllers/multicluster/internal/local_cluster_manager.go:
```diff
@@ -0,0 +1,85 @@
+	// Access to read is likely not needed as the reconciler already is setup to notify changes
+	common.CommonArea
+
+	Start() error
```
please add comments about Start and Stop as well

aravindakidambi replied:
Done, but they are not very meaningful at this point

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+	connected bool
+
+	// if this is the elected leader
+	isElectedLeader bool
```
is this necessary as part of remote cluster manager?

aravindakidambi replied:
Not used right now, so I will remove it.

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+	if !found {
+		return nil, nil, fmt.Errorf("ca.crt data not found in secret %v", secretName)
+	} else {
+	}
```
TODO?

aravindakidambi replied:
Not needed, thanks, I removed it now

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+		if localClusterMemberAnnounce.Annotations == nil {
+			localClusterMemberAnnounce.Annotations = make(map[string]string)
+		}
+		localClusterMemberAnnounce.Annotations["touch-ts"] = time.Now().String()
```
what is the meaning of this annotation? i know this is for timestamps but good to add a note about it.

also use a constant and perhaps a more descriptive name

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+		}
+	} else {
+		// create happens first before a leader can be elected, when the create is successful
+		// it marks the connectivity status to leader election can occur
```
```suggestion
		// it marks the connectivity status to leader election can occur.
```

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+	} else {
+		// create happens first before a leader can be elected, when the create is successful
+		// it marks the connectivity status to leader election can occur
+		// Therefore the first create will not populate the leader cluster id
```
```suggestion
		// Therefore the first create will not populate the leader cluster id.
```

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+		// it marks the connectivity status to leader election can occur
+		// Therefore the first create will not populate the leader cluster id
+		localClusterMemberAnnounce.ClusterID = string(r.remoteClusterManager.GetLocalClusterID())
+		localClusterMemberAnnounce.Name = "member-announce-from-" + string(r.remoteClusterManager.GetLocalClusterID())
```
better to use constants

abhiraut replied:
may also be good to turn this in to a util function to retrieve Name based on input ID

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+		localClusterMemberAnnounce.Namespace = r.Namespace
+		localClusterMemberAnnounce.ClusterSetID = string(r.ClusterSetID)
+		if err := r.Create(ctx, &localClusterMemberAnnounce, &client.CreateOptions{}); err != nil {
+			r.log.Error(err, "Failed to update member announce")
```
```suggestion
			r.log.Error(err, "Failed to create member announce")
```

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,313 @@
+
+/**
+ * Once connected to the remote cluster, the start method run a timer
+ * on a go routing to periodically write MemberAnnounce CRD into the remote
```
```suggestion
 * on a go routine to periodically write MemberAnnounce CRD into the remote
```

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster_manager.go:
```diff
@@ -0,0 +1,154 @@
+package internal
+
+import (
+	"antrea.io/antrea/multicluster/controllers/multicluster/common"
```
imports are not sorted.. same with other files

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster_manager.go:
```diff
@@ -0,0 +1,154 @@
+	Start() error
+	// Stop RemoteClusterManager
+	Stop() error
+	// AddRemoteCluster adds a member cluster to RemoteClusterManager.
```
```suggestion
	// AddRemoteCluster adds a remote cluster to RemoteClusterManager.
```
?

jianjuns commented on multicluster/controllers/multicluster/common/common_area.go:
```diff
@@ -0,0 +1,20 @@
+)
+
+// Access to common area of a cluster
+// Inbound mode:
```
Again suggest not to mention inbound mode, which is confusing to readers without contexts. If you really want to talk about it, at least add comments somewhere to describe what is inbound mode.

jianjuns commented on multicluster/config/samples/multicluster_v1alpha1_clusterset.yaml:
```diff
@@ -1,7 +1,17 @@
+        # serviceAccount: "" # fill this
+    members:
+      - clusterID: test-member-1
+        # secret: "" # This is NA for inbound mode
```
Ditto.

aravindakidambi replied:
Thanks for pointing out

jianjuns commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -18,38 +18,92 @@ package multicluster
+	clusterID        common.ClusterID
+
+	// These fields are only applicable on leader cluster
+	InboundLeaderMgr internal.LocalClusterManager
```
Why not just name it "LeaderMgr"?

aravindakidambi replied:
I changed it to LocalClusterManager now, I missed changing it after renaming the class, hope this is ok?

jianjuns commented on multicluster/controllers/multicluster/internal/remote_cluster_manager.go:
```diff
@@ -0,0 +1,154 @@
+	GetLocalClusterID() common.ClusterID
+}
+
+// The manager manages all leaders defined in the cluster set and manages common area access into them
```
Nit - better to add "." at the end of sentences, esp. when your comments have multiple sentences.

aravindakidambi replied:
Sure, thank you for pointing it out

jianjuns commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,105 @@
+ *               |
+ *            Member1
+ * The member cluster connects to both leaders and performs Member Announce
+ * to announce itself. And all resources from the Member is written into
```
is -> are

jianjuns commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,105 @@
+ * cluster and is considered "connected" if it can successfully write so.
+ * After leader election is done, the result is also used to update the
+ * spec of the MemberAnnounce CRD during the next periodic write into
+ * the leader's CommonArea, so the leader knows it is the elected-leader
```
CommonArea -> Common Area?

jianjuns commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,105 @@
+ * After leader election is done, the result is also used to update the
+ * spec of the MemberAnnounce CRD during the next periodic write into
+ * the leader's CommonArea, so the leader knows it is the elected-leader
+ * of this member and therefore compute necessary resources for it.
```
compute -> computes

jianjuns commented on multicluster/controllers/multicluster/internal/leader_elector.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+/**
+ * Lets consider a ClusterSet with 2 leader clusters and 1 member cluster
```
We discussed this topic before. I am not very convinced we should go active/active leaders, and each leader computes imported resources for its electors (BTW, if each member just chooses its own leader independently, should it be even called "election"?). If we go this way, does it mean all leaders might duplicate the same computation for different member clusters? Do you see a possibility multiple members can share imported resources (for example, if one Service should be imported by 10 members, do we compute 10 times, and create 10 imported resources)?

aravindakidambi replied:
Actually maybe its just my bad, maybe my comment is incorrect. Leader is not computing import resource per member, there is only one import resource that all members will import from the elected-leader. I changed the comment now. I think the only reason leader needs to know which member it is leader for is I think visibility. Im thinking maybe we need to do some cleanup action.  E.g. if at some point, L1 was the leader but M1 changed to L2 as leader, if M1 had written some status for import-resources with either success/error, then maybe L1 should do a cleanup job to clear those.

jianjuns replied:
Ah, that sounds much better. I still feel we might consider a model with only one active elected leader, and all members just write/watch from that leader. But I am fine to start with the current active/active design too, if we already have most code for that. Yes, no matter which way we go we should handle leader change, and re-sync and cleanup after that.

aravindakidambi replied:
Sure, thank you. I will take it up in a future patch.

abhiraut commented:
could you add a detailed description to the PR? so we know what we merged just by looking at the PR message @aravindakidambi

abhiraut commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -18,38 +18,94 @@ package multicluster
+		if !errors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		// if errors.IsNotFound(err)
```
remove this when you address TODOs

abhiraut commented on multicluster/controllers/multicluster/internal/remote_cluster.go:
```diff
@@ -0,0 +1,312 @@
+	// TODO: resource monitoring methods needs to be defined.
+}
+
+// This implements the commonArea interface and allows local cluster to read/write into
```
```suggestion
// This implements the common area interface and allows local cluster to read/write into
```

abhiraut reviewed (approved):
TODOs remain, which should be addressed when the PR is opened up for feature branch

abhiraut commented:
Merging it now. But we plan to follow up for TODOs which will eventually be used to open the PR for feature branch.

luolanzone commented on multicluster/controllers/multicluster/clusterset_controller.go:
```diff
@@ -18,38 +18,94 @@ package multicluster
+	// if update,
+	//    if leader - nothing to do in inbound mode
+	//    if member - if leaders have changed handle accordingly, else nothing to do.
+	if !r.IsLeader {
```
looks like we only support dedicated leader cluster here, I think we should support a situation that a cluster act as both a leader and a member.
